### PR TITLE
Move docs to python modules

### PR DESCRIPTION
**Pull request by thomaspinder**
This PR converts the existing notebooks used in our documentation in Python modules. This allows for easier version control. When building the documentation, each of the modules will be executed with the outputs converted into HTML. 

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [x] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Right now all documentation is held in `.ipynb`. This has the unintended consequence that each time the notebook is run, hundreds of lines of code are staged for commit as the notebook's outputs will change.

Issue Number: #13 

## What is the new behavior?

Documentation is now in static Python modules that are executed and converted to HTML each time the documentation is built.

## Other information

We may need to adjust some settings is some notebooks take a very long time to compile. Long term, we may wish to explore cache-ing the notebooks too for faster build times.

### CODE REVIEW

File: .readthedocs.yaml [hunk 1]
```diff
@@ -14,4 +14,8 @@ build:
 # Build documentation in the docs/ directory with Sphinx
 sphinx:
   configuration: docs/source/conf.py
-  fail_on_warning: false
\ No newline at end of file
+  fail_on_warning: false
+
+python:
+  install:
+    - requirements: docs/requirements.txt
```
Comment (gianlucadetommaso) [inline .readthedocs.yaml:21]: I guess this will need change.

File: docs/README.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+## Build the documentation
+
+The documentation for Fortuna is comprised of a series of notebooks. To serve these locally, there are two steps that need to be taken that we outline below.
+
+### Prerequisites
+
+To build the documentation, first install Fortuna and it's dependencies by following the [installation instructions](https://github.com/awslabs/fortuna#installation). Next, install the documentation requirements through
+```bash
+pip install -r docs/requirements.txt
```
Comment (gianlucadetommaso) [inline docs/README.md:13]: I think this should change. We should try to do everything through poetry.

File: docs/requirements.txt [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# Notebook dependencies
```
Comment (gianlucadetommaso) [inline docs/requirements.txt:1]: I'd prefer to have no `requirements.txt` files whatsoever. We should be able to handle this with poetry.

File: docs/source/conf.py [hunk 4]
```diff
@@ -44,8 +45,14 @@
 }
 autoclass_content = "both"
 
-nbsphinx_allow_errors = True
-nbsphinx_execute = "never"
+nb_execution_mode = "auto"
+nbsphinx_allow_errors = False
```
Comment (gianlucadetommaso) [inline docs/source/conf.py:49]: We might get blocked by some annoying sphinx issues if we set this to False.

File: docs/source/conf.py [hunk 5]
```diff
@@ -44,8 +45,14 @@
...
+nbsphinx_custom_formats = {
+    ".pct.py": ["jupytext.reads", {"fmt": "py:percent"}],
+}
+nbsphinx_execute_arguments = ["--InlineBackend.figure_formats={'svg', 'pdf'}"]
+# If window is narrower than this, input/output prompts are on separate lines:
+nbsphinx_responsive_width = "700px"
```
Comment (gianlucadetommaso) [inline docs/source/conf.py:55]: Could you shortly describe for me what these nbsphinx commands are doing? Thanks!

File: .readthedocs.yaml (hunk 1 above)
Comment (thomaspinder) [reply .readthedocs.yaml:21]: Yes, the original aim was to remove `pip install` commands from inside notebooks. I hadn't realised these requirements could be moved into Poetry.

File: docs/README.md (hunk 2 above)
Comment (thomaspinder) [reply docs/README.md:13]: Agreed.

File: docs/source/conf.py (hunk 4 above)
Comment (thomaspinder) [reply docs/source/conf.py:49]: Potentially. However, it'll also allow us to identify issues within notebooks quickly. We should decide whether this tradeoff is worth it. In my opinion, occasional Sphinx issues are OK if it means we never have notebook issues. Keen to hear your thoughts.

File: docs/source/conf.py (hunk 5 above)
Comment (thomaspinder) [reply docs/source/conf.py:55]: Line 53 renders figures as `svg` files. This is preferable to `png` as the images are vectors so will scale to large displays better. The `pdf` option is there for people exporting the docs to LaTeX or pdfs.
Line 55 will ensure code chunks render nicely on mobiles/tables.

Comment (gianlucadetommaso) [reply docs/source/conf.py:55]: Great, thanks!

File: docs/source/conf.py (hunk 4 above)
Comment (gianlucadetommaso) [reply docs/source/conf.py:49]: Yeah ok, let's try this! We can easily explore this option and see and if setting False become a significant problem or not.

Comment (thomaspinder) [reply docs/source/conf.py:49]: Great! I'll address these comments (unresolved them so I don't forget) and finish this PR next week.

File: docs/README.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+The documentation for Fortuna is comprised of a series of notebooks. To serve these locally, there are two steps that need to be taken that we outline below.
+
+### Prerequisites
+
+To build the documentation, first install Fortuna and it's dependencies by following the [installation instructions](https://github.com/awslabs/fortuna#installation). Next, install the documentation requirements through
+```bash
+poetry install -E docs
+```
+Finally, install `jupytext` through either pip or conda, details can be found [here](https://github.com/mwouts/jupytext#install).
```
Comment (gianlucadetommaso) [inline docs/README.md:15]: How come does this need to be a separate step? Could `jupytext` be included in `docs`?

File: docs/README.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+This will create a corresponding notebook file for each `.pct.py` file that can be opened in Jupyter.
+
+### Building the documentation
+
+From the root directory, documentation can be built by running the following commands:
+
+```bash
+cd docs
+make html
```
Comment (gianlucadetommaso) [inline docs/README.md:33]: I wonder if it is worth making explicit that in order to build the documentation with a poetry environment one would need to run `poetry run make html`, or get into the environment first. What do you think?

File: docs/source/conf.py [hunk 8]
```diff
@@ -14,6 +15,36 @@
...
+            dest_filename = source_filename.replace(examples_source, examples_dest)
+            # If we're skipping examples, put a dummy file in place
+            if os.getenv("SKIP_EXAMPLES"):
+                if dest_filename.endswith("index.rst"):
+                    shutil.copyfile(source_filename, dest_filename)
+                else:
+                    with open(os.path.splitext(dest_filename)[0] + ".rst", "w") as f:
+                        basename = os.path.splitext(os.path.basename(dest_filename))[0]
+                        f.write(f"{basename}\n" + "=" * 80)
+
```
Comment (gianlucadetommaso) [inline docs/source/conf.py:43]: Cool trick, thanks!!

**Review by gianlucadetommaso — COMMENTED** [review]
Great job! Just a couple of comments/questions before merging it in. 😄

File: docs/README.md (hunk 6 above)
Comment (thomaspinder) [reply docs/README.md:15]: Excellent point. Done!

File: docs/README.md (hunk 7 above)
Comment (thomaspinder) [reply docs/README.md:33]: Nice. Poetry is new to me and I didn't know you could do this. Updated.

File: docs/README.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+```bash
+poetry install -E docs
+```
+
+### Notebooks
+
+For easier version control, the notebooks are stored as `.pct.py` files. To convert these to `.ipynb` files, run the following command from the root of the repository:
+
+```bash
+jupytext --to notebook examples/*pct.py
```
Comment (gianlucadetommaso) [inline docs/README.md:21]: If jupytext is installed with poetry, I think we need `poetry run juptext ...`.

File: pyproject.toml [hunk 10]
```diff
@@ -28,12 +28,28 @@ jupyter = {version = "^1.0.0", optional = true}
 
 [tool.poetry.extras]
 docs = ["Sphinx", "sphinx-autodoc-typehints", "pydata-sphinx-theme", "nbsphinx", "nbsphinx-link",
-    "sphinx-gallery", "ipython"]
+    "sphinx-gallery", "ipython", "pandas", "tensorflow-datasets", "xlrd", "openpyxl", "yfinance", 'tabulate']
```
Comment (gianlucadetommaso) [inline pyproject.toml:31]: Should this include jupitext?

**Pull request merged**